# origin: torvalds (BitKeeper)
# cset: 1.1982.161.32 (2.6) key=4244be82bvD-_7wwLkYa0visB12aEw
# URL: http://linux.bkbits.net:8080/linux-2.6/cset@4244be82bvD-_7wwLkYa0visB12aEw
# inclusion: upstream
# descrition: [PATCH] isofs: Handle corupted rock-ridge info slightly better
# revision date: Mon, 28 Mar 2005 18:48:57 +0900
#
# S rset: ChangeSet|1.1982.161.31..1.1982.161.32
# I rset: fs/isofs/rock.c|1.22..1.22.1.1
#
# Key:
# S: Skipped  ChangeSet file only
# O: Original Followed by Updated
# U: Updated  Included with updated range of versions
# I: Included Included verbatim
# E: Excluded Excluded on request from user
# D: Deleted  Manually deleted by subsequent user edit
# R: Revised  Manually revised by subsequent user edit
#
#
# This is a BitKeeper generated diff -Nru style patch.
#
# ChangeSet
#   2005/03/25 17:44:34-08:00 torvalds@osdl.org 
#   [PATCH] isofs: Handle corupted rock-ridge info slightly better
#   
#   Michal Zalewski <lcamtuf@dione.ids.pl> discovers range checking flaws in
#   iso9660 filesystem.
#   
#   http://marc.theaimsgroup.com/?l=bugtraq&m=111110067304783&w=2
#   
#   CAN-2005-0815 is assigned to this issue.
#   
#   From: Linus Torvalds <torvalds@osdl.org>
#   
#   isofs: Handle corupted rock-ridge info slightly better.
#   
#   Keyword here being 'slightly'. The code is a mess.
#   
#   Signed-off-by: Chris Wright <chrisw@osdl.org>
# 
# fs/isofs/rock.c
#   2005/03/25 15:25:50-08:00 torvalds@osdl.org +14 -7
#   isofs: Handle corupted rock-ridge info slightly better
# 
#
===== fs/isofs/rock.c 1.22 vs 1.22.1.1 =====
--- 1.22/fs/isofs/rock.c	2004-09-10 17:47:00 +09:00
+++ 1.22.1.1/fs/isofs/rock.c	2005-03-26 08:25:50 +09:00
@@ -53,6 +53,7 @@
   if(LEN & 1) LEN++;						\
   CHR = ((unsigned char *) DE) + LEN;				\
   LEN = *((unsigned char *) DE) - LEN;                          \
+  if (LEN<0) LEN=0;                                             \
   if (ISOFS_SB(inode->i_sb)->s_rock_offset!=-1)                \
   {                                                             \
      LEN-=ISOFS_SB(inode->i_sb)->s_rock_offset;                \
@@ -103,12 +104,13 @@ int get_rock_ridge_filename(struct iso_d
     struct rock_ridge * rr;
     int sig;
     
-    while (len > 1){ /* There may be one byte for padding somewhere */
+    while (len > 2){ /* There may be one byte for padding somewhere */
       rr = (struct rock_ridge *) chr;
-      if (rr->len == 0) goto out; /* Something got screwed up here */
+      if (rr->len < 3) goto out; /* Something got screwed up here */
       sig = isonum_721(chr);
       chr += rr->len; 
       len -= rr->len;
+      if (len < 0) goto out;	/* corrupted isofs */
 
       switch(sig){
       case SIG('R','R'):
@@ -122,6 +124,7 @@ int get_rock_ridge_filename(struct iso_d
 	break;
       case SIG('N','M'):
 	if (truncate) break;
+	if (rr->len < 5) break;
         /*
 	 * If the flags are 2 or 4, this indicates '.' or '..'.
 	 * We don't want to do anything with this, because it
@@ -186,12 +189,13 @@ parse_rock_ridge_inode_internal(struct i
     struct rock_ridge * rr;
     int rootflag;
     
-    while (len > 1){ /* There may be one byte for padding somewhere */
+    while (len > 2){ /* There may be one byte for padding somewhere */
       rr = (struct rock_ridge *) chr;
-      if (rr->len == 0) goto out; /* Something got screwed up here */
+      if (rr->len < 3) goto out; /* Something got screwed up here */
       sig = isonum_721(chr);
       chr += rr->len; 
       len -= rr->len;
+      if (len < 0) goto out;	/* corrupted isofs */
       
       switch(sig){
 #ifndef CONFIG_ZISOFS		/* No flag for SF or ZF */
@@ -462,7 +466,7 @@ static int rock_ridge_symlink_readpage(s
 	struct rock_ridge *rr;
 
 	if (!ISOFS_SB(inode->i_sb)->s_rock)
-		panic ("Cannot have symlink with high sierra variant of iso filesystem\n");
+		goto error;
 
 	block = ei->i_iget5_block;
 	lock_kernel();
@@ -487,13 +491,15 @@ static int rock_ridge_symlink_readpage(s
 	SETUP_ROCK_RIDGE(raw_inode, chr, len);
 
       repeat:
-	while (len > 1) { /* There may be one byte for padding somewhere */
+	while (len > 2) { /* There may be one byte for padding somewhere */
 		rr = (struct rock_ridge *) chr;
-		if (rr->len == 0)
+		if (rr->len < 3)
 			goto out;	/* Something got screwed up here */
 		sig = isonum_721(chr);
 		chr += rr->len;
 		len -= rr->len;
+		if (len < 0)
+			goto out;	/* corrupted isofs */
 
 		switch (sig) {
 		case SIG('R', 'R'):
@@ -543,6 +549,7 @@ static int rock_ridge_symlink_readpage(s
       fail:
 	brelse(bh);
 	unlock_kernel();
+      error:
 	SetPageError(page);
 	kunmap(page);
 	unlock_page(page);
